processing/src/main/java/io/druid/query/extraction/LookupReferencesManager.java

```diff
@@ -0,0 +1,151 @@
+ *  with the License. You may obtain a copy of the License at
+ *
+ *  http://www.apache.org/licenses/LICENSE-2.0
+ *
+ *  Unless required by applicable law or agreed to in writing,
+ *  software distributed under the License is distributed on an
+ *  "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ *  KIND, either express or implied. See the License for the
+ *  specific language governing permissions and limitations
+ *  under the License.
+ * /
+ */
+
+package io.druid.query.extraction;
+
+
+import com.google.common.collect.ImmutableMap;
+import com.metamx.common.ISE;
+import com.metamx.common.lifecycle.LifecycleStart;
+import com.metamx.common.lifecycle.LifecycleStop;
+import com.metamx.common.logger.Logger;
+import io.druid.guice.ManageLifecycle;
+
+import java.io.IOException;
+import java.util.concurrent.ConcurrentHashMap;
+import java.util.concurrent.ConcurrentMap;
+import java.util.concurrent.atomic.AtomicBoolean;
+
+/**
+ * This class provide a basic {@link LookupExtractor} references manager.
+ * It allows basic operations fetching, listing, adding and deleting of {@link LookupExtractor} objects
+ * It is be used by queries to fetch the lookup reference.
+ * It is used by Lookup configuration manager to add/remove or list lookups configuration via HTTP.
+ */
+
+@ManageLifecycle
+public class LookupReferencesManager
+{
+  private static final Logger LOGGER = new Logger(LookupReferencesManager.class);
+  private final ConcurrentMap<String, LookupExtractor> lookupMap = new ConcurrentHashMap();
+  private final Object lock = new Object();
+  private final AtomicBoolean started = new AtomicBoolean(false);
+
+  @LifecycleStart
+  public void start()
+  {
+    synchronized (lock) {
+      if (!started.getAndSet(true)) {
+        LOGGER.info("Started lookup references manager");
+      }
+    }
+  }
+
+  @LifecycleStop
+  public void stop()
+  {
+    synchronized (lock) {
+      if (started.getAndSet(false)) {
+        for (String lookupName : lookupMap.keySet()) {
+          remove(lookupName);
+        }
+      }
+    }
+  }
+
+  /**
+   * @param lookupName      namespace of the lookup object
+   * @param lookupExtractor {@link LookupExtractor} implementation reference.
+   *
+   * @return true if the lookup is added otherwise false.
+   *
+   * @throws ISE If the manager is closed.
+   */
+  public boolean addLookup(String lookupName, final LookupExtractor lookupExtractor) throws ISE
+  {
+    synchronized (lock) {
+      assertStarted();
+      return (null == lookupMap.putIfAbsent(lookupName, lookupExtractor));
+    }
+  }
+
+  /**
+   * @param lookupName namespace of {@link LookupExtractor} to delete from the reference registry.
+   *                   this function does call the cleaning method {@link LookupExtractor#close()}
+   *
+   * @return true if it is removed
+   */
+  public boolean remove(String lookupName)
+  {
+    final LookupExtractor lookupExtractor = lookupMap.remove(lookupName);
```

Comment:
I agree with @b-slim . The general contract on close() ensures idempotence. It would be worth noting in the LookupExtractor interface that close() is expected to be thread safe and idempotent.